cmd/kubeadm/app/util/etcd.go

```diff
@@ -18,34 +18,61 @@ package util
 
 import (
 	"context"
-	"github.com/coreos/etcd/clientv3"
+	"crypto/tls"
 	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/pkg/transport"
 )
 
-// EtcdCluster is an interface to get etcd cluster related information
-type EtcdCluster interface {
-	GetEtcdClusterStatus() (*clientv3.StatusResponse, error)
+// EtcdClusterInterrogator is an interface to get etcd cluster related information
+type EtcdClusterInterrogator interface {
```

Comment:
I'm of the opinion we should limit the dep expansion and move to wrap and isolate by just adding a Versions f(n). 

Most of the external processing is just string comparison, so we should limit it to that and not proliferate parsing etcd deps, b/c they have been known to shift over time. 

envelop(letter) wrap it.

Reply:
Getting versions is just a byproduct of gathering status. 

Yes, I have added a check to ensure that there is no version mismatch on the etcd nodes (ie. in case etcd was partially upgraded), but the primary objective is to connect to each etcd node and ensure that it is available before attempting an upgrade. This was the original functionality that was in place, with the exception that it only verified etcd at 127.0.0.1.

Reply:
I understand what you are doing.  I'm invoking campsite rules.  Every time we tread in an area, instead of just amending the simple change we need to fix a specific issue, we should strive to make it better than what it was before, otherwise, we end up with a brittle patch-ball.  

The abstraction leakage is an example.  Lets not do that, and instead create a utility function that gets the version(s) b/c that's all your code was looking for anyway.